be/src/exec/pipeline/scan/morsel.cpp

```diff
@@ -42,5 +48,186 @@ StatusOr<MorselPtr> FixedMorselQueue::try_get() {
     }
 }
 
+std::vector<TInternalScanRange*> PhysicalSplitMorselQueue::olap_scan_ranges() const {
+    return _convert_morsels_to_olap_scan_ranges(_morsels);
+}
+
+void PhysicalSplitMorselQueue::set_key_ranges(const std::vector<std::unique_ptr<OlapScanRange>>& key_ranges) {
+    for (const auto& key_range : key_ranges) {
+        if (key_range->begin_scan_range.size() == 1 && key_range->begin_scan_range.get_value(0) == NEGATIVE_INFINITY) {
+            continue;
+        }
+
+        _range_start_op = key_range->begin_include ? "ge" : "gt";
+        _range_end_op = key_range->end_include ? "le" : "lt";
+
+        _range_start_key.emplace_back(key_range->begin_scan_range);
+        _range_end_key.emplace_back(key_range->end_scan_range);
+    }
+}
+
+StatusOr<MorselPtr> PhysicalSplitMorselQueue::try_get() {
+    DCHECK(!_tablets.empty());
+    DCHECK(!_tablet_rowsets.empty());
+    DCHECK_EQ(_tablets.size(), _tablet_rowsets.size());
+
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    if (_tablet_idx >= _tablets.size()) {
+        return nullptr;
+    }
+
+    // When it hasn't initialized any segment,
+    // or _segment_idx exceeds the segments of the current rowset,
+    // or current segment is empty or finished,
+    // we should pick up the next segment and init it.
+    while (!_has_init_any_segment || _cur_segment() == nullptr || _cur_segment()->num_rows() == 0 ||
+           !_segment_range_iter.has_more()) {
+        if (!_next_segment()) {
+            return nullptr;
+        }
+
+        RETURN_IF_ERROR(_init_segment());
```

Comment:
`PhysicalSplitMorselQueue::try_get` is invoked in pipeline execution thread and `_init_segment` may have blocking operation(such as load_index), will it take a negative effect?

Reply:
Yes, `_init_segment` does some i/o operations, which may block the exec thread. However, these i/o operations can not be avoided, since splitting tablet depends on the index and meta of tablets.

Because the default session variable `enable_tablet_internal_parallel` is false for now, maybe in the next PR, I could move the operations about picking up morsels and committing i/o tasks from the exec threads to the i/o threads. Make ScanOperator only pull chunk from the buffer.